### PR TITLE
Private/bryan black lite 1

### PR DESCRIPTION
**Pull request by ManOnTheMountainTech**
1. Straightened out framework and driver owned WDFREQUEST's.
2. Refactored code to make more understandable and according to the NT group coding standards.
3. Code does need cleanup. Lots of residual experiments lying around.
4. Postponed intentionally setting the taillight to black until after the driver has initialized via a work item.
5. SEND_AND_FORGET is supposed to be used for just sending IRP's down the stack. Based on this article, https://www.osr.com/nt-insider/2016-issue1/tips-for-using-io-targets/, it may be calling IoSkipCurrentIrpStackLocation which means certain doom if anything in the request is modified. So there is a fail-safe path for the unmodified request and a separate path for any requests that modify anything. At least until all paths are proven.
6. Work item sends the taillight request down using a synchronous call, as I think it is easier to initially validate.

### CODE REVIEW

**Comment by forderud** [comment]
Thanks for starting to tackle this problem. The PR contain a lot of simultaneous changes though, which makes reviewing hard. Do you think it could be possible to somehow split up the changes into multiple smaller PR's that can be reviewed and merged more independently?

File: TailLight/TailLight.vcxproj [hunk 1]
```diff
@@ -33,6 +33,9 @@
     <DriverType>KMDF</DriverType>
     <PlatformToolset>WindowsKernelModeDriver10.0</PlatformToolset>
     <ConfigurationType>Driver</ConfigurationType>
+    <KMDF_VERSION_MINOR>33</KMDF_VERSION_MINOR>
+    <KMDF_MINIMUM_VERSION_REQUIRED>25</KMDF_MINIMUM_VERSION_REQUIRED>
+    <ALLOW_DATE_TIME>1</ALLOW_DATE_TIME>
```
Comment (forderud) [inline TailLight/TailLight.vcxproj:38]: Are you sure that you really need to change these project settings?

File: TailLight/vfeature.cpp [hunk 2]
```diff
@@ -199,37 +202,38 @@ Routine Description:
         return STATUS_BUFFER_TOO_SMALL;
     }
 
-    TailLightReport* packet = nullptr;
-    NTSTATUS status = WdfRequestRetrieveInputBuffer(Request, sizeof(TailLightReport), (void**)&packet, NULL);
-    if (!NT_SUCCESS(status) || !packet) {
-        KdPrint(("TailLight: WdfRequestRetrieveInputBuffer failed 0x%x, packet=0x%p\n", status, packet));
+    status = WdfRequestRetrieveInputBuffer(Request, sizeof(TailLightReport), (void**)&report, NULL);
```
Comment (forderud) [inline TailLight/vfeature.cpp:205]: Why are you provided a pointer to a self-allocated buffer (`report`) when calling `WdfRequestRetrieveInputBuffer`?

My understanding is that `WdfRequestRetrieveInputBuffer` will give you a direct pointer to  the existing buffer associated with the input request. This is at least how the existing code functions. It's then no point in allocating a buffer in advance with `ExAllocatePool2` as you're doing here.

**Review by forderud — COMMENTED** [review]
I've now tested these changes. Unfortunately, the tail-light does _not_ seem to go black when connecting the mouse after installing this driver.

File: TailLight/TailLight.vcxproj [hunk 3]
```diff
@@ -87,6 +90,9 @@ copy ..\IntelliMouse.ps1 $(OutDir)</Command>
       <PreprocessorDefinitions>%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <ExceptionHandling>
       </ExceptionHandling>
+      <ProgramDataBaseFileName>$(OutDir)$(TargetName).pdb</ProgramDataBaseFileName>
+      <AssemblerOutput>All</AssemblerOutput>
+      <PreprocessToFile>false</PreprocessToFile>
```
Comment (forderud) [inline TailLight/TailLight.vcxproj:95]: What are you trying to achieve with these project settings? PDB files are already being generated for both debug & release builds.

File: TailLight/device.cpp [hunk 4]
```diff
@@ -130,28 +395,129 @@ Routine Description:
 --*/
 {
     UNREFERENCED_PARAMETER(OutputBufferLength);
-    UNREFERENCED_PARAMETER(InputBufferLength);
-
-    //KdPrint(("TailLight: EvtIoDeviceControl (IoControlCode=0x%x, InputBufferLength=%Iu)\n", IoControlCode, InputBufferLength));
 
+    RequestFlags incomingRequestFlags = (RequestFlags)WdfRequestGetInformation(Request);
```
Comment (forderud) [inline TailLight/device.cpp:399]: What kind of information are you trying to get in `incomingRequestFlags`?

I don't see WdfRequestSetInformation or WdfRequestCompleteWithInformation being called anywhere  here to populate the "information" parameter in advance.

**Comment by ManOnTheMountainTech** [comment]
Because I didn't understand the API. It's already fixed. Just need to check in.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by ManOnTheMountainTech** [comment]
There are some bugs in older versions. Is backward compatibility an issue?

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by forderud** [comment]
> There are some bugs in older versions. Is backward compatibility an issue?

Not quite sure what you're thinking about here, but the HID communication protocol is dictated by Microsoft and we therefore need to stay compatible with it. I also want to keep the existing `TailLightDeviceInformation` WMI class unchanged for user-space compatibility, unless there's good reasons for changing it.

As for driver implementation then I of course want bugs to be fixed, but don't want bugfixes to be bundled with enhancements. Therefore, please submit bugfixes in separate pull-requests that we merge _before_ the black tail-light enhancement.

**Comment by ManOnTheMountainTech** [comment]
This week  I will be separating out PR's in logical chunks with documentation.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by forderud** [comment]
Is this pull request still being worked on? If not, then I suggest closing it. If yes, then please resolve the merge conflict and rebase it on top of master. I also suggest shrinking it down so to reduce the number of lines changed.

**Pull request closed without merging**